# Multiple conformation analysis #

The PBs assignment of several conformations (generated by molecular dynamics or whatever method you want) of the same protein can be analyzed statistically. Such analysis is a two-steps process:

1. `PBcount.py` computes the frequency of PBs at each position along the amino acid sequence.
2. `PBstat.py` generates frequency and logo plots, and estimates something similar to entropy called the equivalent number of PBs (_Neq_). 

*Neq* is calculated as follows:

![Neq](img/Neq.jpg)

Where *fx* is the probability of PB *x*. *Neq* quantifies the average number of PBs at a given position in the protein sequence. A *Neq* value of 1 indicates that only one type of PB is observed, while a value of 16 is equivalent to a random distribution. 

eventually, another option is to [cluster structures](#cluster) protein conformations based on their PB sequences.

----

## `PBcount.py`

For this example, first perform PB assignment (with `PBassign.py`):

    ./PBassign.py -p ./demo2/md_traj.pdb -o md_traj

Output:

    1 PDB file(s) to process
    read PB definitions: 16 PBs x 8 angles 
    demo2/md_traj.pdb
    PBs assigned for demo2/md_traj.pdb | model 0
    PBs assigned for demo2/md_traj.pdb | model 1
    PBs assigned for demo2/md_traj.pdb | model 2
    PBs assigned for demo2/md_traj.pdb | model 3
    [snip]
    PBs assigned for demo2/md_traj.pdb | model 221
    PBs assigned for demo2/md_traj.pdb | model 222
    PBs assigned for demo2/md_traj.pdb | model 223
    PBs assigned for demo2/md_traj.pdb | model 224
    wrote md_traj.PB.fasta

Then compute PBs frequency:

    ./PBcount.py -f md_traj.PB.fasta -o md_traj

Output:

    read 225 sequences in md_traj.PB.fasta
    wrote md_traj.PB.count

Content of `md_traj.PB.count`:

             a     b     c     d     e     f     g     h     i     j     k     l     m     n     o     p
    1        0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0
    2        0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0
    3        0     0     0     0     0    98     0     0     0     0   127     0     0     0     0     0
    4        0     0     0     0     0     0     0     0     0     0   101   124     0     0     0     0
    5        0     0     0     0     0     0     0     0     0     0     1   118   106     0     0     0
    6        0     0     0     0     0     0     0     0     0     0     0     0   224     1     0     0
    7        0     0     0     0     0     0   158     0     0     0     0     0    47    20     0     0
    8        0     0     0     0     0     0     0     0     0     0     0     0    46     0   179     0
    9        0     0     0     0     0     0     0     0     0    76     2     0    48     0     0    99
    [snip]
    687      3     0     2   124     0    92     2     0     0     0     0     2     0     0     0     0
    688      0    63    35    26    50    23    23     0     0     0     5     0     0     0     0     0
    689      0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0
    690      0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0


### Usage

    usage: PBcount.py [-h] -f F -o O [--first-residue FIRST_RESIDUE]

    Compute PB frequency along protein sequence.

    optional arguments:
      -h, --help            show this help message and exit
      -f F                  name(s) of the PBs file (in fasta format)
      -o O                  name for results
      --first-residue FIRST_RESIDUE
                            define first residue number (1 by default)


### `-f` option

can be used several times:

    ./PBcount.py -f demo2/md_traj_1.PB.fasta -f demo2/md_traj_2.PB.fasta -f demo2/md_traj_3.PB.fasta -o test

Output:

    read 90 sequences in demo2/md_traj_1.PB.fasta
    read 90 sequences in demo2/md_traj_2.PB.fasta
    read 90 sequences in demo2/md_traj_3.PB.fasta
    wrote test.PB.count

### `--first-residue` option

By default, the number of the first residue is 1, this option allows to adjust the number associated to the first residue (and to the followings automaticaly).

    ./PBcount.py --residue-shift 5 -f demo2/md_traj_1.PB.fasta -o test1

Output:

    read 90 sequences in demo2/md_traj_1.PB.fasta
    first residue will be numbered 5
    wrote test1.PB.count

Content of `test1.PB.count`:

             a     b     c     d     e     f     g     h     i     j     k     l     m     n     o     p
    5        0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0
    6        0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0
    7        0     0     0     0     0    90     0     0     0     0     0     0     0     0     0     0
    8        0     0     0     0     0     1     0     0     0     0    89     0     0     0     0     0
    9        0    89     0     0     0     0     0     0     0     0     0     1     0     0     0     0
    10       0     0    86     0     0     3     0     0     0     0     0     0     1     0     0     0
    11       0     0     1     0     0     7     4     0     0     0     0     0     0    78     0     0
    12       0    48     0     0     0     0     0     9     0     0     0     7     0     0    26     0
    [snip]


----

## `PBstat.py`

generates frequency and logo plots and estimates *Neq*. 

To generate map or Neq plot, PBxplore requires R to be installed and present in the user PATH.

To generate Weblogo-like represations, PBxplore requires R to be installed and present in the user PATH.

Example:

    ./PBstat.py -f md_traj.PB.count --map --neq --logo -o md_traj

Output:

    Index of first residue is: 1
    wrote md_traj.PB.map.png

    wrote md_traj.PB.Neq
    wrote md_traj.PB.Neq.png

    wrote md_traj.PB.logo.pdf


### Usage

    usage: PBstat.py [-h] -f F -o O [--map] [--neq] [--logo]
                     [--residue-min RESIDUE_MIN] [--residue-max RESIDUE_MAX]

    Statistical analysis and graphical representations of PBs.

    optional arguments:
      -h, --help            show this help message and exit
      -f F                  name of file that contains PBs frequency (count)
      -o O                  name for results
      --map                 generate map of the distribution of PBs along protein
                            sequence
      --neq                 compute Neq and generate Neq plot along protein
                            sequence
      --logo                generate logo representation of PBs frequency along
                            protein sequence
      --residue-min RESIDUE_MIN
                            defines lower bound of residue frame
      --residue-max RESIDUE_MAX
                            defines upper bound of residue frame

### `--map` option

generates map of the distribution of PBs along protein sequence.

Example:

    ./PBstat.py -f demo2/md_traj.PB.count --map -o md_traj

Output:

    wrote md_traj.PB.map.png

Graph:

![distribution of PBs](img/md_traj.PB.map.jpg "distribution of PBs")

### `--neq` option

computes *Neq* and generates *Neq* plot along protein sequence.

Example:

    ./PBstat.py -f demo2/md_traj.PB.count --neq -o md_traj

Output:

    wrote md_traj.PB.Neq
    wrote md_traj.PB.Neq.png

Content of `md_traj.PB.Neq`:

    resid       Neq 
    1          1.00 
    2          1.00 
    3          1.99 
    4          2.00 
    5          2.06 
    6          1.04 
    7          2.21 
    8          1.67 
    [snip]

Graph:

![Neq versus residue number](img/md_traj.PB.Neq.jpg "Neq versus residue number")


### `--logo` option

generates WebLogo-like representation of PBs frequency along protein sequence. This option requires Weblogo [1] to be installed.

Example:

    ./PBstat.py -f demo2/md_traj.PB.count --logo -o md_traj

Output:

    wrote md_traj.PB.logo.pdf

Graph:

![logo representation of PBs frequency](img/md_traj.PB.logo.jpg "logo representation of PBs frequency")



### `--residue-min` and `--residue-max` options

These options define the lower and upper bound of residue frame. 

    ./PBstat.py -f md_traj.PB.count --map --neq --logo -o md_traj --residue-min 15 --residue-max 35

Output:

    wrote md_traj.PB.map.15-35.png

    wrote md_traj.PB.Neq.15-35
    wrote md_traj.PB.Neq.15-35.png

    wrote md_traj.PB.logo.15-35.pdf

PBs distribution:

![PBs distribution with residue frame](img/md_traj.PB.Neq.frame.jpg)

Neq versus residue number:

![Neq versus residue number with residue frame](img/md_traj.PB.Neq.frame.jpg "Neq versus residue number with residue frame")

Logo representation of PBs frequency:

![Logo representation of PBs frequency](img/md_traj.PB.logo.frame.jpg "Logo representation of PBs frequency with residue frame")


## <a name="cluster"></a>Clustering of protein structures

Once converted to PB sequences, conformations of a same protein structure car be clustered based on PB similarities with `PBclust.py`. 

Example: 

    ./PBclust.py -f demo2/beta3_IEGF12.PB.fasta -o beta3_IEGF12

Output:

    read 190 sequences in demo2/beta3_IEGF12.PB.fasta
    read substitution matrix
    Building distance matrix
    100%
    wrote beta3_IEGF12.PB.dist
    R clustering: OK
    cluster   1:     9 sequences (  4%)
    cluster   2:    16 sequences (  8%)
    cluster   3:    48 sequences ( 25%)
    cluster   4:    52 sequences ( 27%)
    cluster   5:    65 sequences ( 34%)
    wrote beta3_IEGF12.PB.clust

Cluster 5 is the biggest cluster with 34% of conformations.


`beta3_IEGF12.PB.dist` contains the matrix distance between all PB sequences.


Content of `md_traj_1.PB.clust` (clustering results):

    SEQ_CLU  "beta3_IEGF12.pdb | model 0"  1 
    SEQ_CLU  "beta3_IEGF12.pdb | model 1"  1 
    SEQ_CLU  "beta3_IEGF12.pdb | model 2"  1 
    [snip]
    ...
    [snip]
    SEQ_CLU  "beta3_IEGF12.pdb | model 26"  2 
    SEQ_CLU  "beta3_IEGF12.pdb | model 27"  3 
    SEQ_CLU  "beta3_IEGF12.pdb | model 28"  3 
    [snip]
    ...
    [snip]
    SEQ_CLU  "beta3_IEGF12.pdb | model 188"  5 
    SEQ_CLU  "beta3_IEGF12.pdb | model 189"  5 
    MED_CLU  "beta3_IEGF12.pdb | model 6"  1 
    MED_CLU  "beta3_IEGF12.pdb | model 18"  2 
    MED_CLU  "beta3_IEGF12.pdb | model 48"  3 
    MED_CLU  "beta3_IEGF12.pdb | model 98"  4 
    MED_CLU  "beta3_IEGF12.pdb | model 157"  5 


### Usage

    usage: PBclust.py [-h] -f F -o O [--clusters CLUSTERS_NB] [--compare]

    Cluster protein structures based on their PB sequences.

    optional arguments:
      -h, --help            show this help message and exit
      -f F                  name(s) of the PBs file (in fasta format)
      -o O                  name for results
      --clusters CLUSTERS_NB
                            number of wanted clusters
      --compare             compare the first sequence versus all others


### `--clusers` option

defines the number of wnated clusters (5 is the default)

Example:

    ./PBclust.py -f demo2/beta3_IEGF12.PB.fasta -o beta3_IEGF12 --clusters 3

Output:

    read 190 sequences in demo2/beta3_IEGF12.PB.fasta
    read substitution matrix
    Building distance matrix
    100%
    wrote beta3_IEGF12.PB.dist
    R clustering: OK
    cluster   1:    25 sequences ( 13%)
    cluster   2:    48 sequences ( 25%)
    cluster   3:   117 sequences ( 61%)
    wrote beta3_IEGF12.PB.clust

### `--compare` option

compares, position by position, the first sequence found in the fasta file against all others. The substitution matric between PBs is converted to a score between O (identical) and 9 (different).

Example:

    ./PBclust.py -f demo2/beta3_IEGF12.PB.fasta -o beta3_IEGF12 --compare

Output:

    read 190 sequences in demo2/beta3_IEGF12.PB.fasta
    read substitution matrix
    Normalized substitution matrix (between 0 and 9)
    [[0 3 2 3 4 3 3 4 2 3 5 3 5 4 3 3]
     [3 0 3 3 3 4 3 2 2 3 3 2 5 3 3 2]
     [2 3 0 3 4 3 2 4 3 4 5 5 5 4 3 2]
     [3 3 3 0 2 3 4 4 3 3 5 5 9 6 5 4]
     [4 3 4 2 0 2 2 2 4 3 3 4 7 4 5 5]
     [3 4 3 3 2 0 3 3 4 2 3 3 5 5 4 5]
     [3 3 2 4 2 3 0 3 3 3 4 3 3 2 2 1]
     [4 2 4 4 2 3 3 0 3 1 2 3 5 4 2 4]
     [2 2 3 3 4 4 3 3 0 2 2 2 5 3 3 2]
     [3 3 4 3 3 2 3 1 2 0 2 2 4 4 3 3]
     [5 3 5 5 3 3 4 2 2 2 0 3 3 3 4 4]
     [3 2 5 5 4 3 3 3 2 2 3 0 3 2 2 4]
     [5 5 5 9 7 5 3 5 5 4 3 3 0 2 3 3]
     [4 3 4 6 4 5 2 4 3 4 3 2 2 0 2 2]
     [3 3 3 5 5 4 2 2 3 3 4 2 3 2 0 2]
     [3 2 2 4 5 5 1 4 2 3 4 4 3 2 2 0]]
    Compare first sequence (beta3_IEGF12.pdb | model 0) with others
    wrote beta3_IEGF12.PB.compare.fasta

Content of `beta3_IEGF12.PB.compare.fasta`:

    >beta3_IEGF12.pdb | model 0 vs beta3_IEGF12.pdb | model 1
    000000200000000020000000000000000000000000200030000300003204
    20240030012200000000000000200
    >beta3_IEGF12.pdb | model 0 vs beta3_IEGF12.pdb | model 2
    000000200000020000000003200000000000000000200000300300000004
    20040430013200000330000000200
    [snip]
    ...
    [snip]
    >beta3_IEGF12.pdb | model 0 vs beta3_IEGF12.pdb | model 187
    003355030000033432330000000000000000000333200030000300000004
    20230000012203333035030000200
    >beta3_IEGF12.pdb | model 0 vs beta3_IEGF12.pdb | model 188
    003355030000023432330000000000000000000333200030303000000004
    20230000012203333035030000200
    >beta3_IEGF12.pdb | model 0 vs beta3_IEGF12.pdb | model 189
    003355030030023432000000000000004000000333200030304000000004
    20230000012203333035030000200


# References

[1] G. E. Crooks, G. Hon, J.-M. Chandonia, and S. E. Brenner. WebLogo: A Sequence Logo Generator. *Genome Research* **14**: 1188–90 (2004)
doi:10.1101/gr.849004 http://weblogo.threeplusone.com/
